File: node/src/prompts.ts
import * as fs from "node:fs";
import * as path from "node:path";
import * as url from "node:url";

let resolvedDir: string | null = null;
const tplCache = new Map<string, string>();

function isPromptsDir(p: string): boolean {
  try {
    return fs.statSync(p).isDirectory();
  } catch {
    return false;
  }
}

function walkUp(start: string, attempted: string[]): string | null {
  let cur = start;
  for (let i = 0; i < 10; i++) {
    const cand = path.join(cur, "prompts");
    attempted.push(cand);
    if (isPromptsDir(cand)) return cand;
    const parent = path.dirname(cur);
    if (parent === cur) break;
    cur = parent;
  }
  return null;
}

export function resolveDir(): string {
  if (resolvedDir !== null) return resolvedDir;
  const attempted: string[] = [];
  const env = process.env.CODEX_WEB_PROMPTS;
  if (env) {
    attempted.push(env);
    if (isPromptsDir(env)) {
      resolvedDir = env;
      return resolvedDir;
    }
  }
  // From the executing script's directory.
  try {
    const here = path.dirname(url.fileURLToPath(import.meta.url));
    const found = walkUp(here, attempted);
    if (found) {
      resolvedDir = found;
      return resolvedDir;
    }
  } catch {
    /* ignore */
  }
  // From cwd.
  try {
    const found = walkUp(process.cwd(), attempted);
    if (found) {
      resolvedDir = found;
      return resolvedDir;
    }
  } catch {
    /* ignore */
  }
  throw new Error(`prompts directory not found; tried: ${attempted.join(", ")}`);
}

export function render(toolName: string, vars: Record<string, string>): string {
  let tpl = tplCache.get(toolName);
  if (tpl === undefined) {
    const dir = resolveDir();
    tpl = fs.readFileSync(path.join(dir, `${toolName}.md`), "utf8");
    tplCache.set(toolName, tpl);
  }
  let out = tpl;
  for (const [k, v] of Object.entries(vars)) {
    out = out.split(`{${k}}`).join(v);
  }
  return out;
}

File: fsharp/src/CodexWebMcp.FSharp/Prompts.fs
module CodexWebMcp.FSharp.Prompts

open System
open System.Collections.Concurrent
open System.IO
open System.Reflection

let private lockObj = obj ()
let mutable private dirValue : string option = None
let private tplCache = ConcurrentDictionary<string, string>()

let private isPromptsDir (p: string) =
    try Directory.Exists(p) with _ -> false

let private walkUp (start: string) (attempted: ResizeArray<string>) : string option =
    let mutable cur = start
    let mutable result = None
    let mutable stop = false
    let mutable i = 0
    while not stop && i < 10 && result.IsNone do
        let cand = Path.Combine(cur, "prompts")
        attempted.Add(cand)
        if isPromptsDir cand then
            result <- Some cand
        else
            let parent = Path.GetDirectoryName(cur)
            if String.IsNullOrEmpty(parent) || parent = cur then
                stop <- true
            else
                cur <- parent
        i <- i + 1
    result

let resolveDir () : string =
    lock lockObj (fun () ->
        match dirValue with
        | Some v -> v
        | None ->
            let attempted = ResizeArray<string>()
            let envVal = Environment.GetEnvironmentVariable("CODEX_WEB_PROMPTS")
            let envFound =
                if not (String.IsNullOrEmpty envVal) then
                    attempted.Add(envVal)
                    if isPromptsDir envVal then Some envVal else None
                else None
            let resolved =
                match envFound with
                | Some v -> Some v
                | None ->
                    let exeDir =
                        try Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)
                        with _ -> ""
                    let fromExe =
                        if not (String.IsNullOrEmpty exeDir) then walkUp exeDir attempted
                        else None
                    match fromExe with
                    | Some _ -> fromExe
                    | None ->
                        try walkUp (Directory.GetCurrentDirectory()) attempted
                        with _ -> None
            match resolved with
            | Some v ->
                dirValue <- Some v
                v
            | None ->
                let joined = String.Join(", ", attempted)
                raise (FileNotFoundException(sprintf "prompts directory not found; tried: %s" joined))
    )

let render (toolName: string) (vars: Map<string, string>) : string =
    let tpl =
        match tplCache.TryGetValue(toolName) with
        | true, v -> v
        | false, _ ->
            let dir = resolveDir ()
            let path = Path.Combine(dir, toolName + ".md")
            let content = File.ReadAllText(path)
            tplCache.TryAdd(toolName, content) |> ignore
            content
    let mutable out = tpl
    for KeyValue(k, v) in vars do
        out <- out.Replace("{" + k + "}", v)
    out
